tensorflow_lite_support/ios/task/audio/core/sources/TFLRingBuffer.m

```diff
@@ -0,0 +1,118 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "tensorflow_lite_support/ios/task/audio/core/sources/TFLRingBuffer.h"
+#import "tensorflow_lite_support/ios/sources/TFLCommon.h"
+#import "tensorflow_lite_support/ios/sources/TFLCommonUtils.h"
+
+@implementation TFLRingBuffer {
+  NSInteger _nextIndex;
+  TFLFloatBuffer *_buffer;
+  NSInteger _size;
+}
+
+- (instancetype)initWithBufferSize:(NSInteger)size {
+  self = [self init];
+  if (self) {
+    _buffer = [[TFLFloatBuffer alloc] initWithSize:size];
+    _size = size;
+  }
+  return self;
+}
+
+- (BOOL)loadBuffer:(TFLFloatBuffer *)sourceBuffer
+            offset:(NSUInteger)offset
+              size:(NSUInteger)size
+             error:(NSError **)error {
+  NSInteger sizeToCopy = size;
+  NSInteger newOffset = offset;
+
+  if (offset + size > sourceBuffer.size) {
+    [TFLCommonUtils
+        createCustomError:error
+                 withCode:TFLSupportErrorCodeInvalidArgumentError
+              description:@"offset + size exceeds the maximum size of the source buffer."];
+    return NO;
+  }
+
+  // Length is greater than buffer size, then modify size and offset to
+  // keep most recent data in the sourceBuffer.
+  if (size >= _buffer.size) {
+    sizeToCopy = _buffer.size;
+    newOffset = offset + (size - _buffer.size);
+  }
+
+  // If the new nextIndex + sizeToCopy is smaller than the size of the ring buffer directly
+  // copy all elements to the end of the ring buffer.
+  if (_nextIndex + sizeToCopy < _buffer.size) {
+    memcpy(_buffer.data + _nextIndex, sourceBuffer.data + newOffset, sizeof(float) * sizeToCopy);
+  } else {
+    // If
+    NSInteger endChunkSize = _buffer.size - _nextIndex;
+    memcpy(_buffer.data + _nextIndex, sourceBuffer.data + newOffset, sizeof(float) * endChunkSize);
+
+    NSInteger startChunkSize = sizeToCopy - endChunkSize;
+    memcpy(_buffer.data, sourceBuffer.data + newOffset + endChunkSize,
+           sizeof(float) * startChunkSize);
+  }
+
+  _nextIndex = (_nextIndex + sizeToCopy) % _buffer.size;
+
+  return YES;
+}
+
+- (TFLFloatBuffer *)floatBuffer {
+  return [self floatBufferWithOffset:0 size:[self size]];
+}
+
+- (nullable TFLFloatBuffer *)floatBufferWithOffset:(NSUInteger)offset size:(NSUInteger)size {
+  if (offset + size > _buffer.size) {
+    return nil;
+  }
+
+  // TODO: Investigate if allocating is needed everytime or a global alloc of self.floatBuffer and
+  // copying of elements each time and returning self.floatBuffer would suffice.
+  TFLFloatBuffer *bufferToReturn = [[TFLFloatBuffer alloc] initWithSize:size];
```

Comment:
It's necessary to alloc every time because the docstring says that this method returns a **copy** of the internal buffer in the correct order.